### PR TITLE
Datasources init container

### PR DESCRIPTION
**Pull request by swtch1**
#### What this PR does / why we need it:
The current version of the chart assumes that datasources, like dashboards, can be added to the filesystem and Grafana will recognize them at any time.  While this is how the documentation reads, this is not true in practice.  Grafana issue [12878](https://github.com/grafana/grafana/issues/12878) shows that the Grafana team actually considers datasources configuration rather than content and should not be dynamically reloaded.  This makes the sidecar method confusing when adding new datasource doesn't correctly add them to the dashboard, and harmful when the race condition between the configmap parsing code and Grafana load does not load all or any datasources.

Here I have moved the sidecar method to an initContainer.  This will pull the datasources down before Grafana is loaded ensuring they always make into the configuration, and this is more explicit about the fact that datasources cannot be reloaded on the fly.

#### Which issue this PR fixes
Works around Grafana issue [12878](https://github.com/grafana/grafana/issues/12878).

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

@zanhsieh 
@rtluckie 

Signed-off-by: Joshua Thornton <thornton.joshua@gmail.com>

### CODE REVIEW

File: stable/grafana/README.md
Comment (KierranM) [inline stable/grafana/README.md]: According to the description this should be:
```suggestion
| `initContainer.datasources.enabled`| Enabled the cluster wide search for datasources and adds/updates/deletes them in grafana once on pod init |`false`       |
```

File: stable/grafana/README.md
Comment (KierranM) [inline stable/grafana/README.md]: Same deal, `datasources` rather than `dashboards`
```suggestion
| `initContainer.datasources.enabled`| Label that config maps with datasources should have to be added | `false`                               |
```

**Review by KierranM — CHANGES_REQUESTED** [review]
Hit this today, thanks for opening a PR. I noticed a few docs changes that are needed

**Comment by zanhsieh** [comment]
@swtch1 Do you mind to resolve DCO first?

File: stable/grafana/templates/deployment.yaml
Comment (meringu) [inline stable/grafana/templates/deployment.yaml]: Can this resources configuration please be added to the init container too?

Comment (swtch1) [reply stable/grafana/templates/deployment.yaml]: The resource configuration has been added to the pull request.  A bit confused by the automatic "pull request successfully merged" without seeing it end up in the master branch, but glad to get it moving forward.

**Comment by RaymondArias** [comment]
whats the status on this pr? I see its approved but not merged

**Comment by tim-beatson** [comment]
I too am wondering what the status of this is? It would be ideal for our use case if this could be merged.

**Comment by swtch1** [comment]
Hey @meringu, do you mind reviewing this again.  It looks like it's ready to be merged.

**Comment by meringu** [comment]
What happened to the commits in this PR? I'm building from my fork that has the commits that were here, If the commits we're to make there way back in the PR I'd be happy to add my approval. I'm also not a maintainer by the way.

**Comment by swtch1** [comment]
Yeah.. I had to do some --force actions to get all of the commits signed.  It appears the changes are still there.

**Comment by tim-beatson** [comment]
The PR has 0 commits

**Comment by m0rganic** [comment]
I'm not sure what the hold up on this PR is but I took at look at @swtch1's forked repo and it worked! 

I had the exact same idea earlier today but stopped when I realized kiwigrid's k8s-sidecar wouldn't run as an initContainer because it long polls the server for configmaps and never completes. Since grafana currently doesn't support changes to provisioned datasources except after startup it doesn't really make sense to use k8s-sidecar for that.

I ran into issues with the race condition and loading of datasources on startup all day long.
 
👍 thanks @swtch1

**Comment by tim-beatson** [comment]
I agree, would be great if he could push the changes in his forked repo up into this PR so we can approve it

**Comment by chrisob** [comment]
@swtch1 It looks like you have the incorrect branch of your fork to be merged in this PR.  

Can you either change the branch to be merged in this PR from `datasources_init_container` to your fork's `master`, or bring the changes you've made to your `master` into your `datasources_init_container` branch?

Otherwise, this PR will show no changes between `helm/charts/master` and `swtch1/charts/datasources_init_container`.

**Comment by xuxinkun** [comment]
This issue also draws my attention. But the content of the pr disappeared. So, I try to fix it on my way and I make this PR. https://github.com/helm/charts/pull/9842
Pls help review.

**Pull request closed without merging**
